### DISCUSSION TITLE
regex filter for lsp_references

### DISCUSSION THREAD

**Discussion by serranomorante** [ibhagwan/fzf-lua#1210] [Q&A] [answered]
Thank you for [this](https://github.com/ibhagwan/fzf-lua/commit/56ae880d7949c45e6efddccddcca4e2bf1f2c0a7) 

I just tried the `regex_filter` but Lua patterns are just extremely extremely limiting 😢
Vim search patterns are a better matching tool (although I don't know about performance)
I also tried with `fzf_opts = { ['--query'] = ...}` on `lsp_references` but that will use a fuzzy matcher, not regex 😮‍💨 .

I want to generate a list of regex "matchers" for several languages, that will exclude the import statements from the `lsp_references` result.

Something like:
```lua
local exclude_imports = {
  python = {"regex_pattern_1", ...}, -- could match anything but "from main import main" or "import main"
  javascript = {"regex_pattern_1", ...}, -- could match anything but "import React from 'react'" or "const react = require('react')"
  etc, etc
}
```
I **don't think** fzf-lua should worry about implementing a new config option just for this (something like `include_imports...`)

Are there any other available options to use regex on the lsp_references?

Maybe any of the [advance](https://github.com/ibhagwan/fzf-lua/wiki/Advanced) entries could help on this? I'm willing to dig further and maybe create my own lsp_references just for my config.

Thanks!

**Comment by ibhagwan** [comment]
> I just tried the regex_filter but Lua patterns are just extremely extremely limiting
> I don't think fzf-lua should worry about implementing a new config option just for this

We can have `regex_filter` accept a function so you can filter the items as you please, the only minor issue would be the ability to display the custom filter in the header (can be displayed as `<function>`).

**Reply by serranomorante** [reply]
It make sense to me! As long as this is not a maintainability burden for you. Thank you so much :heart:

**Comment by ibhagwan** [comment, marked as answer]
@serranomorante, see https://github.com/ibhagwan/fzf-lua/commit/a9e5755721c0f24cd1635ea4ff181df5b1bd5dd4

Added some improvements including the below:
```lua
:FzfLua lsp_references regex_filter={"import.*from.*",exclude=true}
-- equal to:
:lua require("fzf-lua").lsp_finder({ regex_filter = {"import.*from.*", exclude=true} })
```

You can also define `regex_filter` as a function, when it returns `false` the item will be filtered out:
```lua
-- Explore `ctx` and `item` with `vim.print` so you can decide on your filtering logic
-- `ctx` contains buffer number and so you can determine `filetype`, etc
:lua require("fzf-lua").lsp_finder({regex_filter=function(e,ctx) vim.print(ctx); return true end})
```

When a function is used the header will display `<function>` , can be disabled with `header=false`:
![image](https://github.com/ibhagwan/fzf-lua/assets/59988195/f66dedda-ecaf-4a99-b3e8-d29602e980c0)

**Reply by serranomorante** [reply]
This is so beautiful. Tested it, works like charm! Thank you so much.

Now I can just do something like this:
```lua
M.javascript_regex_filter = function(item)
  if item.text:match("^import%s.*from.*") then return false end
  return true
end
M.python_regex_filter = function(item)
  if item.text:match("^from%s.*import.*") then return false end
  if item.text:match("^import%s.*") then return false end
  return true
end
```

Very convenient :heart: :heart: :heart:

**Reply by ibhagwan** [reply]
Very nice indeed!
